### PR TITLE
Grpc.Tools package has MSBuild/XBuild magic to generate C# source code

### PR DESCRIPTION
**Pull request by ananthonline**
The changes in this pull request contain changes that enable users to automagically generate C# source code for .proto files included in their project.

All they have to do is install the Nuget package (and possibly reload their solution). Now the "Build Action" for proto files will contain a new item called "Proto". If selected, these files generate C# in the obj/$(Configuration) folder and include them as part of the build.

This should be completely cross-platform, I've tested it on Windows + VS2013 and OSX + Xamarin Studio.

@jtattermusch, this was the reason I wanted binaries for all platforms in this nuget package. Thanks for helping with that request!

### CODE REVIEW

File: src/csharp/Grpc.Tools.CSharpGenerator/Program.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                if (buf != IntPtr.Zero)
+                    Marshal.FreeHGlobal(buf);
+            }
+            return false;
+        }        
+        static void Main(string[] args)
+        {
+            try
+            {
+                var projectDir = args[0];
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools.CSharpGenerator/Program.cs]: The platform detection code already exists in Grpc.Core.Internal.PlatformApis, having a slightly different version of the same code here feels odd. On the other hand, we can't really make code generator project depend on Grpc.Core.  Anyway, I feel like a cleaner solution is needed here.

File: src/csharp/Grpc.Tools.CSharpGenerator/Program.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                var outputDir = Path.Combine(projectDir, intermediateOutputDir);
+
+                var inputFiles = from protofile in args.Skip(2).TakeWhile(s => true)
+                                 select Path.Combine(projectDir, protofile);
+                foreach (var f in inputFiles)
+                    Console.WriteLine("Going to process: {0}", f);
+
+                foreach (var inputFile in inputFiles)
+                {
+                    var protocArguments = string.Format("-I{0} --csharp_out {1} --grpc_out {1} --plugin=protoc-gen-grpc={2} {3}", protosIncludeDir, outputDir, grpcCSharpPluginPath, inputFile);
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools.CSharpGenerator/Program.cs]: Does this have a way to specify custom protoc options/custom flags?

File: src/csharp/Grpc.Tools/Grpc.Tools.nuspec [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    <requireLicenseAcceptance>false</requireLicenseAcceptance>
+    <releaseNotes>Release $version$</releaseNotes>
+    <copyright>Copyright 2015, Google Inc.</copyright>
+    <tags>gRPC RPC Protocol HTTP/2</tags>
+	<dependencies>
+	<dependency id="Grpc" version="$version$" />
+	</dependencies>
+  </metadata>
+  <files>
+    <file src="build\Grpc.Tools.targets" target="build\Grpc.Tools.targets" />
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/Grpc.Tools.nuspec]: Grpc.Tools is a solution-level nuget package. Is Grpc.Tools.targets something that's allowed in solution-level packages? Or does adding a .targets file make the nuget package become a project-level package?

File: src/csharp/build_packages.bat [hunk 4]
```diff
@@ -8,7 +8,7 @@ set PROTOBUF_VERSION=3.0.0-beta2
 set VERSION_WITH_BETA=%VERSION%-beta
 
 @rem Adjust the location of nuget.exe
-set NUGET=C:\nuget\nuget.exe
```
Comment (jtattermusch) [inline src/csharp/build_packages.bat]: I think this would break our automated package building. please revert the change.

File: src/csharp/Grpc.Tools.CSharpGenerator/Grpc.Tools.CSharpGenerator.csproj [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  </ItemGroup>
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
+  <Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
+    <PropertyGroup>
+      <ErrorText>This project references NuGet package(s) that are missing on this computer. Enable NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
+    </PropertyGroup>
+    <Error Condition="!Exists('$(SolutionDir)\.nuget\NuGet.targets')" Text="$([System.String]::Format('$(ErrorText)', '$(SolutionDir)\.nuget\NuGet.targets'))" />
+  </Target>
+  <!-- To modify your build process, add your task inside one of the targets below and uncomment it.
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools.CSharpGenerator/Grpc.Tools.CSharpGenerator.csproj]: Why this comment?

**Comment by jtattermusch** [comment]
this is ok to test

**Comment by jtattermusch** [comment]
I generally like the idea of this contribution - thanks for the pull request. There's a couple of problems though:
-- we need to have a plan how to support this for  CoreCLR/.NET (msbuild XML files is not how the new .NET projects are built).
-- some way of customizing the codegen args will probably be necessary.
-- Technically this problem is something we should solve in in google/protobuf project first and then apply the same logic for gRPC.

Btw, we already have some related issues: https://github.com/grpc/grpc/issues/4805 and https://github.com/google/protobuf/issues/1193. It is probably worth reading through, as it identifies some of the challenges.

Also, CC @jskeet.

File: src/csharp/Grpc.Tools.CSharpGenerator/Program.cs (hunk 1 above)
Comment (ananthonline) [inline src/csharp/Grpc.Tools.CSharpGenerator/Program.cs]: I think it might be ok to have this depend on Grpc.Core - because this is a build time only dependency. Unfortunately this class is internal so I can't use it as-is. Would you feel better if I used some other nuget package that included this check? Again, note that this a build-time dependency only.

File: src/csharp/Grpc.Tools.CSharpGenerator/Program.cs (hunk 2 above)
Comment (ananthonline) [inline src/csharp/Grpc.Tools.CSharpGenerator/Program.cs]: At this point, no. How would you want to wire these in? I assumed most of the common knobs and dials were provided by proto [options](https://developers.google.com/protocol-buffers/docs/proto3#options). Could you give me an example of the kinds of customizations you were thinking of?

File: src/csharp/Grpc.Tools/Grpc.Tools.nuspec (hunk 3 above)
Comment (ananthonline) [inline src/csharp/Grpc.Tools/Grpc.Tools.nuspec]: You're correct. As pointed out (here)[https://docs.nuget.org/create/creating-and-publishing-a-package#creating-a-solution-level-package], adding anything in the lib, build or content folders makes this a project-level package.

Could you help me understand why having Grpc.Tools as a solution level package is desirable?

File: src/csharp/build_packages.bat (hunk 4 above)
Comment (ananthonline) [inline src/csharp/build_packages.bat]: Will do. I assumed that nuget would be available in $(SolutionDir)/.nuget/nuget.exe, I will revert this change.

File: src/csharp/Grpc.Tools.CSharpGenerator/Grpc.Tools.CSharpGenerator.csproj (hunk 5 above)
Comment (ananthonline) [inline src/csharp/Grpc.Tools.CSharpGenerator/Grpc.Tools.CSharpGenerator.csproj]: Ah, this is just the default project generated by Visual Studio. I didn't edit this in any way. Would you like me to get rid of it?

File: src/csharp/Grpc.Tools/Grpc.Tools.nuspec (hunk 3 above)
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/Grpc.Tools.nuspec]: I don't think it's any more desirable for it to be a solution level package than a project level package. It just happens that currently it is a solution level package and changing that require a manual intervention from everyone who has already installed the package at the solution level. Depending on the circumstances tat might be fine, I just wanted to point that out.

File: src/csharp/Grpc.Tools.CSharpGenerator/Program.cs (hunk 1 above)
Comment (jtattermusch) [inline src/csharp/Grpc.Tools.CSharpGenerator/Program.cs]: At this point you can probably just copy-and-paste the PlatformApi class and add a big comment that this is actually what Grpc.Core already implements. Then we are at least duplicating the very same code and we can figure out the rest in the next step.

File: src/csharp/Grpc.Tools.CSharpGenerator/Program.cs (hunk 2 above)
Comment (jtattermusch) [inline src/csharp/Grpc.Tools.CSharpGenerator/Program.cs]: C# protobuf currently supports e.g. the base_namespace cmdline option.

File: src/csharp/Grpc.Tools.CSharpGenerator/Grpc.Tools.CSharpGenerator.csproj (hunk 5 above)
Comment (jtattermusch) [inline src/csharp/Grpc.Tools.CSharpGenerator/Grpc.Tools.CSharpGenerator.csproj]: I'd like to VS project to look as similar as possible to the other C# projects. But we can figure that out as the last step, after we resolve all the conceptual problems (like how to support .NET core).
Btw, have you considered adding just a wrapper script instead of CSharpGenerator.exe written in C#? Basically all we need here is something that wraps the platform-specific executables in an a portable fashion.

File: src/csharp/Grpc.Tools/Grpc.Tools.nuspec (hunk 3 above)
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/Grpc.Tools.nuspec]: Here's a thought: 

Instead of trying to add this functionality to Grpc.Tools (and possibly breaking users), we could as a first step add a contrib package (e.g. `Grpc.Tools.MSBuildIntegration` ) that would contain your changes. Than users could opt-in to use the msbuild integration. Later, once that solution proves itself useful, we will be in better position to decide how exactly we'd like to integrate it with the other official Grpc packages.

Comment (ananthonline) [inline src/csharp/Grpc.Tools/Grpc.Tools.nuspec]: I'm ok with that approach too, but because Nuget adds a version number to package names, Grpc.Tools.MSBuildIntegration would have to also contain the protoc and grpc_csharp_plugin binaries. This means I'll essentially be replicating Grpc.Tools and adding the build folder (with the targets file) in it.

If that sounds good to you, I can have a pull request ready in a bit. As for being able to customize the command line, I'd like to add this in a subsequent release (I'm trying to think up clean ways to do this).

Do let me know, thanks!

File: src/csharp/Grpc.Tools.CSharpGenerator/Program.cs (hunk 2 above)
Comment (ananthonline) [inline src/csharp/Grpc.Tools.CSharpGenerator/Program.cs]: As a first step, I can support this using the "Custom Tool namespace" metadata on that item. 

Later, I can extend it to be generic by option name. See this example of custom metadata [here](https://msdn.microsoft.com/en-us/library/ms171474.aspx). Consumers that want to use this advanced feature will have to hand-edit their project files and add the metadata in.

Does this sound good?

File: src/csharp/Grpc.Tools.CSharpGenerator/Grpc.Tools.CSharpGenerator.csproj (hunk 5 above)
Comment (ananthonline) [inline src/csharp/Grpc.Tools.CSharpGenerator/Grpc.Tools.CSharpGenerator.csproj]: I could use a script, but instead of writing and maintaining 2 (3 code paths) scripts, it seemed easier to write a portable executable. As it stands, I can also change the logic inside the generator to be arbitrarily complex if needed. I thought the risks were small and the rewards many with this approach.

File: src/csharp/Grpc.Tools.CSharpGenerator/Program.cs (hunk 1 above)
Comment (ananthonline) [inline src/csharp/Grpc.Tools.CSharpGenerator/Program.cs]: Ok, I can do that.

**Comment by ananthonline** [comment]
Closing this pull request. Will reopen with separate Nuget package called Grpc.Tools.MSBuildIntegration that addresses many of the features requested on this thread.

**Comment by hcoona** [comment]
However, the #6085 died. Do you have any plan to continue the project or add more advanced support to MSBuild?

**Comment by jtattermusch** [comment]
We might revisit that at some point, but for now, we have no plans to work on that for the next few quarters. Contributions are welcome though (but please be aware that this is a tricky problem to solve and it needs to support more than just trivial .proto files - see previous discussion).
